### ISSUE TITLE
Is it possible to get instance segmentation in the pipeline?

### ISSUE THREAD

**Issue by Zailushang211** [helblazer811/ConceptAttention#3]
Hi Alec,
Nice work! Thanks a lot for sharing.
I was just wondering if it is possible to realize instance segmentation in the pipeline while generating an image?

**Comment by helblazer811** [comment]
Hmmm. I suspect this approach would be ill-suited for instance segmentation between objects with the same or a similar class. For example, if you have a picture with two dogs, the patches where the dogs are will likely have similar scores w.r.t a "dog" concept.

**Comment by Zailushang211** [comment]
> Hmmm. I suspect this approach would be ill-suited for instance segmentation between objects with the same or a similar class. For example, if you have a picture with two dogs, the patches where the dogs are will likely have similar scores w.r.t a "dog" concept.

Thanks a lot, it seems so. Then it could be interesting to  figure out that how does the model decides the position of an instance.

**Comment by helblazer811** [comment]
I suspect you will have success if you look at the self attentions of
objects. This can be done in a similar fashion to our approach. Take a
patch on your object of interest and find the other patches with the
highest self attention to it. Perhaps you can also use the output space
vectors to find something like this.

**Comment by Zailushang211** [comment]
Thanks again, I will do some search

**Comment by tunahansalih** [comment]
Instance segmentation would be a valuable addition to the paper but currently ConceptAttention doesn't provide instance-level segmentation maps. A solution for instance-level segmentation would be clustering self-attention maps to separate pixel-level activations of instances.

**Comment by venkatesh-thiru** [comment]
@tunahansalih @helblazer811 
I have been exploring extending Concept Attention for zero-shot instance segmentation in FLUX, specifically to identify and separate multiple people in an image (e.g., "A photo of two people").

### My Current Pipeline:
* **Concept Extration:** Generate the concept map for the concept "face".
<img width="552" height="291" alt="Image" src="https://github.com/user-attachments/assets/1a23c534-e264-41e8-badf-ec86eaaa719f" />
 
* **Clustering:** I apply a threshold and use `scipy.ndimage.label` to identify distinct "islands"(centroid) for each person. 
<img width="416" height="413" alt="Image" src="https://github.com/user-attachments/assets/baa82171-2266-4a3a-9216-210a4632c4a9" />
 
* **Self-Attention Query:** I modified the DoubleStreamBlock to extract the image self-attention. I then query this map using the patch indices from the clusters identified earlier

<img width="543" height="291" alt="Image" src="https://github.com/user-attachments/assets/36613267-46f5-431e-a6d6-36242d8f9a8a" />

### Issue:
Though the concept map successfully finds two distinct clusters, the resulting self-attention maps of these instances show significant overlap as seen in the heatmap above, the attention for "Person A" bleeds heavily into "Person B", resulting in a lack of distinction between the two instances.

### Image Self-Attention:
I currently extract the self attention map like the following

```
# Inside DoubleStreamBlock
d_head = img_q.shape[-1]
scale = 1.0 / math.sqrt(d_head)

# Calculating Q @ K^T
img_self_scores = torch.einsum("b h i d, b h j d -> b h i j", img_q, img_k) * scale
img_self_attn_map = img_self_scores.softmax(dim=-1)

concept_attention_dict['img_self_attn'] = img_self_attn_map.detach().cpu()
```


Do you have any suggestions for sharpening these maps to enforce separation? I am trying to avoid training a dedicated probe if possible.

Is there a recommended ways to aggregate the attention from multiple patches within a cluster to reduce this "ghosting" effect?
